kore/src/Kore/Log/DebugRewriteTrace.hs

```diff
@@ -0,0 +1,267 @@
+ )
+import Data.Yaml (
+    ToJSON,
+    Value (..),
+    array,
+    encode,
+    object,
+    toJSON,
+    (.=),
+ )
+import Kore.Attribute.UniqueId (
+    UniqueId (..),
+ )
+import Kore.Internal.Conditional qualified as Conditional
+import Kore.Internal.MultiOr (
+    MultiOr,
+ )
+import Kore.Internal.Pattern (
+    Pattern,
+ )
+import Kore.Internal.Predicate (
+    unparseWithSort,
+ )
+import Kore.Internal.Substitution (
+    Substitution,
+    assignedTerm,
+    assignedVariable,
+    unwrap,
+ )
+import Kore.Internal.Substitution qualified as Substitution
+import Kore.Internal.TermLike qualified as TermLike
+import Kore.Internal.TermLike.TermLike (
+    TermLike,
+ )
+import Kore.Internal.Variable (
+    VariableName,
+    toVariableName,
+ )
+import Kore.Rewrite.Result qualified as Result
+import Kore.Rewrite.RewritingVariable
+import Kore.Rewrite.RulePattern (
+    UnifyingRuleVariable,
+ )
+import Kore.Rewrite.Step (
+    Results,
+ )
+import Kore.Unparser (
+    Unparse,
+    unparse,
+ )
+import Log
+import Prelude.Kore
+import Pretty (
+    Pretty (..),
+    layoutOneLine,
+    renderText,
+ )
+
+data DebugInitialClaim = DebugInitialClaim UniqueId (TermLike VariableName)
+    deriving stock (Show)
+
+newtype DebugInitialPattern = DebugInitialPattern (TermLike VariableName)
+    deriving stock (Show)
+
+newtype DebugFinalPatterns = DebugFinalPatterns [Pattern VariableName]
+    deriving stock (Show)
+
+data RewriteResult = RewriteResult
+    { ruleId :: UniqueId
+    , substitution :: Substitution VariableName
+    , results :: [Pattern VariableName]
+    }
+    deriving stock (Show)
+
+data DebugRewriteTrace = DebugRewriteTrace
+    { initialPattern :: Pattern VariableName
+    , rewriteResults :: [RewriteResult]
+    , remainders :: [Pattern VariableName]
+    }
+    deriving stock (Show)
+
+instance ToJSON (TermLike VariableName) where
+    toJSON = toJSON . unparseOneLine
+
+instance ToJSON (Pattern VariableName) where
+    toJSON Conditional.Conditional{term, predicate, substitution} =
+        object
+            [ "term" .= term
+            , "constraint" .= encodedConstraint
+            , "substitution" .= substitution
+            ]
+      where
+        sort = TermLike.termLikeSort term
+        encodedConstraint = (renderText $ layoutOneLine $ unparseWithSort sort predicate) :: Text
+
+instance ToJSON (Substitution VariableName) where
+    toJSON substitution =
+        array (encodeKV <$> unwrap substitution)
+      where
+        encodeKV assignment =
+            object
+                [ "key" .= unparseOneLine (assignedVariable assignment)
+                , "value" .= unparseOneLine (assignedTerm assignment)
+                ]
+
+instance ToJSON DebugInitialClaim where
+    toJSON (DebugInitialClaim uniqueId claim) =
+        object
+            [ "task" .= ("reachability" :: Text)
+            , "claim" .= claim
+            , "claim-id" .= maybe Null toJSON (getUniqueId uniqueId)
+            ]
```

Comment:
(related to the above, using `renderX :: X -> Value` instead of implicit `ToJSON` via `(.=)` and using `aeson` not `yaml`)
Maybe worthwhile to define a direct `toEncoding` here for better performance?